pkg/k8s/endpointsynchronizer/cep.go

```diff
@@ -179,73 +182,83 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
 					return nil
 				}
 
+				// Serialize the endpoint into a model. It is compared with the one
+				// from before, only updating on changes.
 				mdl := e.GetCiliumEndpointStatus()
 				if reflect.DeepEqual(mdl, lastMdl) {
 					scopedLog.Debug("Skipping CiliumEndpoint update because it has not changed")
 					return nil
 				}
 
-				cep, err := ciliumClient.CiliumEndpoints(namespace).Get(podName, meta_v1.GetOptions{})
-				switch {
-				// The CEP doesn't exist. We will fall through to the create code below
-				case err != nil && k8serrors.IsNotFound(err):
-					break
-
-				// Delete the CEP on the first ever run. We will fall through to the create code below
-				case firstRun:
-					firstRun = false
-					scopedLog.Debug("Deleting CEP on first run")
+				if needInit {
+					scopedLog.Debug("Deleting CEP during an initialization")
 					err := ciliumClient.CiliumEndpoints(namespace).Delete(podName, &meta_v1.DeleteOptions{})
-					if err != nil {
+					// It's only an error if it exists but something else happened
+					if err != nil && !k8serrors.IsNotFound(err) {
 						scopedLog.WithError(err).Warn("Error deleting CEP")
 						return err
 					}
 
-				// Delete an invalid CEP. We will fall through to the create code below
-				case err != nil && k8serrors.IsInvalid(err):
-					scopedLog.WithError(err).Warn("Invalid CEP during update")
-					err := ciliumClient.CiliumEndpoints(namespace).Delete(podName, &meta_v1.DeleteOptions{})
+					// We can't create localCEP directly, it must come from the k8s server.
+					cep := &cilium_v2.CiliumEndpoint{
+						ObjectMeta: meta_v1.ObjectMeta{
+							Name: podName,
+						},
+						Status: *mdl,
+					}
+					localCEP, err = ciliumClient.CiliumEndpoints(namespace).Create(cep)
 					if err != nil {
-						scopedLog.WithError(err).Warn("Error deleting invalid CEP during update")
+						scopedLog.WithError(err).Error("Cannot create CEP")
 						return err
 					}
 
-				// A real error
-				case err != nil && !k8serrors.IsNotFound(err):
-					scopedLog.WithError(err).Error("Cannot get CEP for update")
-					return err
+					needInit = false
+				}
 
-				// do an update
-				case err == nil:
+				// We have an object to reuse. Update it and push it up.
+				if localCEP != nil {
+					scopedLog.Debug("Updating CEP from local copy")
 					// Update the copy of the cep
-					mdl.DeepCopyInto(&cep.Status)
+					mdl.DeepCopyInto(&localCEP.Status)
 					var err2 error
 					switch {
 					case ciliumUpdateStatusVerConstr.Check(k8sServerVer):
-						_, err2 = ciliumClient.CiliumEndpoints(namespace).UpdateStatus(cep)
+						localCEP, err2 = ciliumClient.CiliumEndpoints(namespace).UpdateStatus(localCEP)
 					default:
-						_, err2 = ciliumClient.CiliumEndpoints(namespace).Update(cep)
+						localCEP, err2 = ciliumClient.CiliumEndpoints(namespace).Update(localCEP)
 					}
 					if err2 != nil {
 						scopedLog.WithError(err2).Error("Cannot update CEP")
+						localCEP = nil
```

Comment:
Is this correct? When `Update()` failed because the revision is outdated, it provides the latest revision ready for retry. Please add comments to describe when exactly Update() fails and how we deal with outdated revisions.

Reply:
You are right, it wasn't correct. I shuffled the code around to handle the revision error case specifically, and to be more clear on the order of operations.